File: examples/word-paragraph-format.ts
/**
 * 例子（Epic-30）：演示 Paragraph.alignment + Paragraph.indent。
 *
 * 跑法：
 *   bun run examples/word-paragraph-format.ts <output.docx>
 */

import { Paragraph, Run, Text, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-paragraph-format <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart?.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  function addPara(
    text: string,
    align?: "left" | "center" | "right",
    indent?: { leftDxa?: number; firstLineDxa?: number },
  ): void {
    const p = new Paragraph();
    const r = new Run();
    const t = new Text();
    t.text = text;
    r.appendChild(t);
    p.appendChild(r);
    if (align !== undefined) p.alignment = align;
    if (indent !== undefined) p.indent = indent;
    append(p);
  }

  addPara("Default left aligned.");
  addPara("This is centered.", "center");
  addPara("This is right aligned.", "right");
  addPara("Indented left 720 dxa (≈ 0.5 inch).", undefined, { leftDxa: 720 });
  addPara("First line indent 480 dxa.", undefined, { firstLineDxa: 480 });
  addPara("Hanging indent 360 dxa with left base 720 dxa.", undefined, {
    leftDxa: 720,
    hangingDxa: 360,
  });

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (6 paragraphs with various format)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordParagraphFormat.cs
// Replica of examples/word-paragraph-format.ts
// Creates a docx with 6 paragraphs demonstrating alignment and indentation.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordParagraphFormat
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        body.Append(MakePara("Default left aligned."));
        body.Append(MakePara("This is centered.", align: "center"));
        body.Append(MakePara("This is right aligned.", align: "right"));
        body.Append(MakePara("Indented left 720 dxa (≈ 0.5 inch).", leftDxa: 720));
        body.Append(MakePara("First line indent 480 dxa.", firstLineDxa: 480));
        body.Append(MakePara("Hanging indent 360 dxa with left base 720 dxa.", leftDxa: 720, hangingDxa: 360));

        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }

    private static Paragraph MakePara(
        string text,
        string? align = null,
        int? leftDxa = null,
        int? firstLineDxa = null,
        int? hangingDxa = null)
    {
        var run = new Run(new Text(text) { Space = SpaceProcessingModeValues.Preserve });
        var para = new Paragraph(run);

        var pPr = new ParagraphProperties();
        bool hasPPr = false;

        if (align != null)
        {
            var jc = align == "center" ? JustificationValues.Center
                   : align == "right"  ? JustificationValues.Right
                   : JustificationValues.Left;
            pPr.Append(new Justification { Val = jc });
            hasPPr = true;
        }

        if (leftDxa != null || firstLineDxa != null || hangingDxa != null)
        {
            var indent = new Indentation();
            if (leftDxa != null) indent.Left = leftDxa.ToString();
            if (firstLineDxa != null) indent.FirstLine = firstLineDxa.ToString();
            if (hangingDxa != null) indent.Hanging = hangingDxa.ToString();
            pPr.Append(indent);
            hasPPr = true;
        }

        if (hasPPr)
            para.PrependChild(pPr);

        return para;
    }
}
